### PR TITLE
Feature/database move

### PR DESCRIPTION
**Pull request by KashmiraDeshpande**

### CODE REVIEW

File: Ginger/DatabaseUnitTest/DataBaseUnitTest.cs [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+            Boolean testconn = db.OpenConnection(param);
+            
+        }
+
+        [TestMethod]
+        public void OpenConnection()
+        {
+            //Arrange
+            Dictionary<string, string> param = new Dictionary<string, string>();
+            param.Add("ConnectionString", @"Provider=Microsoft.ACE.OLEDB.12.0;Data Source= C:\GingerSourceControl\Solutions\Ginger_Regression_Testing\Documents\WOEI\SignUp.accdb;");
```
Comment (yaronweiss) [inline Ginger/DatabaseUnitTest/DataBaseUnitTest.cs]: hard coded c:\ - please change to use testresources

File: Ginger/DatabaseUnitTest/DataBaseUnitTest.cs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+            //Arrange
+            
+            List<string> Tables= null;
+            
+            //Act
+
+             Tables= db.GetTablesList();
+           
+            //Assert
+            Assert.AreEqual(Tables.Count, 2);
```
Comment (yaronweiss) [inline Ginger/DatabaseUnitTest/DataBaseUnitTest.cs]: Add check of the actual table names as expected

File: Ginger/DatabaseUnitTest/DataBaseUnitTest.cs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        {
+            //Arrange
+            List<string> Columns = null;
+            string tablename = "Person";
+            
+            //Act
+            Columns = db.GetTablesColumns(tablename);
+           
+            //Assert
+            Assert.AreEqual(Columns.Count, 9);
```
Comment (yaronweiss) [inline Ginger/DatabaseUnitTest/DataBaseUnitTest.cs]: Add check for all 9 columns expected

File: Ginger/DatabaseUnitTest/DataBaseUnitTest.cs [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        {
+            //Arrange
+            string upadateCommand = "update Person set LName=\"EFGH\" where ID=1";
+            string result = null;
+            
+            //Act
+            result = db.RunUpdateCommand(upadateCommand, false);
+           
+            //Assert
+            Assert.AreEqual(result, "1");
```
Comment (yaronweiss) [inline Ginger/DatabaseUnitTest/DataBaseUnitTest.cs]: add anothe read from the DB to verify it was really upadted

File: Ginger/DatabaseUnitTest/DataBaseUnitTest.cs [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        }
+
+        [TestMethod]
+        public void DBQuery()
+        {
+            //Arrange
+            List<object> result = null;
+            
+            //Act
+            result = db.DBQuery("Select * from Person");
```
Comment (yaronweiss) [inline Ginger/DatabaseUnitTest/DataBaseUnitTest.cs]: DBQuery needs to return a table no List<object> - let's discuss

File: Ginger/DatabaseUnitTest/DataBaseUnitTest.cs [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        public void GetRecordCount()
+        {
+            //Arrange
+             int a = 0;
+            
+            //Act
+                a = db.GetRecordCount("Person");
+            
+            //Assert
+            Assert.AreEqual(a, 2);
```
Comment (yaronweiss) [inline Ginger/DatabaseUnitTest/DataBaseUnitTest.cs]: 2 is expected needs to be first param, then add ex info what we check so if it fails the err will be clear

File: Ginger/DatabaseUnitTest/DataBaseUnitTest.cs [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        }
+
+        [TestCleanup]
+        public void CloseConnection()
```
Comment (yaronweiss) [inline Ginger/DatabaseUnitTest/DataBaseUnitTest.cs]: remove hard coded c:\

File: Ginger/DatabaseUnitTest/UnitTest1.cs [hunk 8]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace MSdatabaseUT
+{
+    [TestClass]
+    public class UnitTest1
```
Comment (yaronweiss) [inline Ginger/DatabaseUnitTest/UnitTest1.cs]: remove this class or change name to meaningful

File: Ginger/GingerCoreNET/Database/IDatabase.cs [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+using System.Collections.Generic;
+using System.Text;
+
+namespace Amdocs.Ginger.CoreNET
+{
+    public interface IDatabase
+    {
+        Boolean OpenConnection(Dictionary<string, string> parameters);
+        void CloseConnection();
+        List<string> GetTablesList(string Name= null);// Keyspace - cassandra ??
```
Comment (yaronweiss) [inline Ginger/GingerCoreNET/Database/IDatabase.cs]: We need to remove the Name will talk about solution

File: Ginger/GingerCoreNET/Database/IDatabase.cs [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+        List<string> GetTablesColumns(string table);
+        string RunUpdateCommand(string updateCmd, bool commit = true);
+        string GetSingleValue(string Table, string Column, string Where);
+        List<object> DBQuery(string Query); //  int? timeout = null : TODO // Return Datatable
```
Comment (yaronweiss) [inline Ginger/GingerCoreNET/Database/IDatabase.cs]: return value we need something like table but for no sql we need to disucss

File: Ginger/MSAccessDB/MSAccessDBCon.cs [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+                Reporter.ToLog(eLogLevel.ERROR, "Failed to close DB Connection", e);
+                throw (e);
+            }
+            finally
+            {
+                conn?.Dispose();
+            }
+        }
+
+        public List<object> DBQuery(string Query)
```
Comment (yaronweiss) [inline Ginger/MSAccessDB/MSAccessDBCon.cs]: We need table return instead

File: Ginger/MSAccessDB/MSAccessDBCon.cs [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+            return ReturnList;
+        }
+       
+        public string GetConnectionString(Dictionary<string,string> parameters)
+        {
+            string connStr = null;
+            bool res;
+            res = false;
+
+            string ConnectionString= parameters.FirstOrDefault(pair => pair.Key == "ConnectionString").Value;
```
Comment (yaronweiss) [inline Ginger/MSAccessDB/MSAccessDBCon.cs]: use Dictionary Get instead of Linq

Comment (yaronweiss) [reply Ginger/MSAccessDB/MSAccessDBCon.cs]: or better to add key value pair needed for below then no need to search

File: Ginger/MSAccessDB/MSAccessDBCon.cs [hunk 13]
```diff
@@ -0,0 +1,352 @@
...
+            {
+                try
+                {
+                    DbCommand command = conn.CreateCommand();
+                    command.CommandText = sql;
+                    command.CommandType = CommandType.Text;
+
+                    // Retrieve the data.
+                    reader = command.ExecuteReader();
+                    while (reader.Read())
```
Comment (yaronweiss) [inline Ginger/MSAccessDB/MSAccessDBCon.cs]: if we know we will read only one then remove the while

File: Ginger/MSAccessDB/MSAccessDBCon.cs [hunk 14]
```diff
@@ -0,0 +1,352 @@
...
+        public bool MakeSureConnectionIsOpen()
+        {
+            Boolean isCoonected = true;
+
+            if ((conn == null) || (conn.State != ConnectionState.Open))
+            {
+                isCoonected = OpenConnection(KeyvalParamatersList);
+            }
+            //make sure that the connection was not refused by the server               
+            TimeSpan timeDiff = DateTime.Now - LastConnectionUsedTime;
```
Comment (yaronweiss) [inline Ginger/MSAccessDB/MSAccessDBCon.cs]: I don't think we need the below check and why 5 mins?

File: Ginger/MSAccessDB/MSAccessDBCon.cs [hunk 15]
```diff
@@ -0,0 +1,352 @@
...
+                            tran = conn.BeginTransaction();
+                            // to Command object for a pending local transaction
+                            command.Connection = conn;
+                            command.Transaction = tran;
+                        }
+                        command.CommandText = updateCmd;
+                        command.CommandType = CommandType.Text;
+
+                        result = command.ExecuteNonQuery().ToString();
+                        if (commit)
```
Comment (yaronweiss) [inline Ginger/MSAccessDB/MSAccessDBCon.cs]: if no commit done then what is the use case?

File: Ginger/DatabaseUnitTest/DataBaseUnitTest.cs (hunk 2 above)
Comment (yaronweiss) [reply Ginger/DatabaseUnitTest/DataBaseUnitTest.cs]: please add the check

File: Ginger/DatabaseUnitTest/DataBaseUnitTest.cs (hunk 4 above)
Comment (yaronweiss) [reply Ginger/DatabaseUnitTest/DataBaseUnitTest.cs]: please handle

File: Ginger/DatabaseUnitTest/DataBaseUnitTest.cs (hunk 5 above)
Comment (yaronweiss) [reply Ginger/DatabaseUnitTest/DataBaseUnitTest.cs]: please handle

File: Ginger/DatabaseUnitTest/DataBaseUnitTest.cs (hunk 6 above)
Comment (yaronweiss) [reply Ginger/DatabaseUnitTest/DataBaseUnitTest.cs]: please handle

File: Ginger/DatabaseUnitTest/DataBaseUnitTest.cs (hunk 2 above)
Comment (KashmiraDeshpande) [reply Ginger/DatabaseUnitTest/DataBaseUnitTest.cs]: added check

File: Ginger/DatabaseUnitTest/DataBaseUnitTest.cs (hunk 4 above)
Comment (KashmiraDeshpande) [reply Ginger/DatabaseUnitTest/DataBaseUnitTest.cs]: added check

File: Ginger/DatabaseUnitTest/DataBaseUnitTest.cs (hunk 6 above)
Comment (KashmiraDeshpande) [reply Ginger/DatabaseUnitTest/DataBaseUnitTest.cs]: added check

File: Ginger/MSAccessDB/MSAccessDBCon.cs (hunk 15 above)
Comment (KashmiraDeshpande) [reply Ginger/MSAccessDB/MSAccessDBCon.cs]: there is a condition in a database action : if user wants to commit he can commit or he can just update

File: Ginger/DatabaseUnitTest/DataBaseUnitTest.cs (hunk 5 above)
Comment (KashmiraDeshpande) [reply Ginger/DatabaseUnitTest/DataBaseUnitTest.cs]: handled

File: Ginger/DatabaseUnitTest/UnitTest1.cs (hunk 8 above)
Comment (KashmiraDeshpande) [reply Ginger/DatabaseUnitTest/UnitTest1.cs]: done

File: Ginger/GingerCoreNET/Database/IDatabase.cs (hunk 10 above)
Comment (KashmiraDeshpande) [reply Ginger/GingerCoreNET/Database/IDatabase.cs]: done

File: Ginger/MSAccessDB/MSAccessDBCon.cs (hunk 11 above)
Comment (KashmiraDeshpande) [reply Ginger/MSAccessDB/MSAccessDBCon.cs]: done

File: Ginger/MSAccessDB/MSAccessDBCon.cs (hunk 13 above)
Comment (KashmiraDeshpande) [reply Ginger/MSAccessDB/MSAccessDBCon.cs]: Done

File: Ginger/MSAccessDB/MSAccessDBCon.cs (hunk 14 above)
Comment (KashmiraDeshpande) [reply Ginger/MSAccessDB/MSAccessDBCon.cs]: removed

**Pull request merged**